lib/Backend/NativeEntryPointData.cpp

```diff
@@ -0,0 +1,568 @@
+#include "JitTransferData.h"
+
+using namespace Js;
+
+NativeEntryPointData::NativeEntryPointData() :
+    runtimeTypeRefs(nullptr), propertyGuardCount(0), propertyGuardWeakRefs(nullptr), polymorphicInlineCacheInfo(nullptr),
+    equivalentTypeCacheCount(0), equivalentTypeCaches(nullptr), registeredEquivalentTypeCacheRef(nullptr),
+    sharedPropertyGuards(nullptr), constructorCaches(nullptr), weakFuncRefSet(nullptr), jitTransferData(nullptr),
+    nativeThrowSpanSequence(nullptr), codeSize(0), nativeAddress(nullptr), thunkAddress(nullptr), validationCookie(nullptr)
+#if PDATA_ENABLED
+    , xdataInfo(nullptr)
+#endif
+#if DBG_DUMP | defined(VTUNE_PROFILING)
+    , nativeOffsetMaps(&HeapAllocator::Instance)
+#endif
+{
+}
+
+JitTransferData* 
+NativeEntryPointData::EnsureJitTransferData(Recycler* recycler)
+{
+    if (this->jitTransferData == nullptr)
+    {
+        this->jitTransferData = RecyclerNew(recycler, JitTransferData);
+    }
+    return this->jitTransferData;
+}
+
+void
+NativeEntryPointData::FreeJitTransferData()
+{
+    JitTransferData * data = this->jitTransferData;
+    if (data != nullptr)
+    {
+        this->jitTransferData = nullptr;
+        data->Cleanup();
+    }
+}
+
+void
+NativeEntryPointData::RecordNativeCode(Js::JavascriptMethod thunkAddress, Js::JavascriptMethod nativeAddress, ptrdiff_t codeSize, void * validationCookie)
+{
+    Assert(this->thunkAddress == nullptr);
+    Assert(this->nativeAddress == nullptr);
+    Assert(this->codeSize == 0);
+    Assert(this->validationCookie == nullptr);
+
+    this->nativeAddress = nativeAddress;
+    this->thunkAddress = thunkAddress;
+    this->codeSize = codeSize;
+    this->validationCookie = validationCookie;
+}
+
+void
+NativeEntryPointData::FreeNativeCode(ScriptContext * scriptContext, bool isShutdown)
+{
+#if PDATA_ENABLED
+    this->CleanupXDataInfo();
+#endif
+
+    // if scriptContext is shutting down, no need to free that native code
+    if (!isShutdown && this->nativeAddress != nullptr)
+    {
+        // In the debugger case, we might call cleanup after the native code gen that
+        // allocated this entry point has already shutdown. In that case, the validation
+        // check below should fail and we should not try to free this entry point
+        // since it's already been freed
+        Assert(this->validationCookie != nullptr);
+        if (this->validationCookie == scriptContext->GetNativeCodeGenerator())
+        {
+            FreeNativeCodeGenAllocation(scriptContext, this->nativeAddress, this->thunkAddress);
+        }
+    }
+
+#ifdef PERF_COUNTERS
+    PERF_COUNTER_SUB(Code, TotalNativeCodeSize, codeSize);
+    PERF_COUNTER_SUB(Code, FunctionNativeCodeSize,codeSize);
+    PERF_COUNTER_SUB(Code, DynamicNativeCodeSize, codeSize);
+#endif
+
+    this->nativeAddress = nullptr;
+    this->thunkAddress = nullptr;
+    this->codeSize = 0;
+}
+
+void
+NativeEntryPointData::SetTJNativeAddress(Js::JavascriptMethod nativeAddress, void * validationCookie)
```

Comment:
Why is this method special? The fields it sets are also set by other methods? Should templatized jit have its own NativeEntryPointData?

Reply:
The method is only used by templated JIT, so I renamed it as such.
Templatize JIT should have it's own Native entry point, but since templated JIT is not enabled, I am not doing that work at this point.